library/src/main/java/com/google/android/exoplayer/extractor/flv/FlvExtractor.java

```diff
@@ -0,0 +1,310 @@
+      input.skipFully(4);
+
+      // Read the tag header from the input.
+      input.readFully(tagHeaderBuffer.data, 0, FLV_TAG_HEADER_SIZE);
+
+      tagHeaderBuffer.setPosition(0);
+      int type = tagHeaderBuffer.readUnsignedByte();
+      int dataSize = tagHeaderBuffer.readUnsignedInt24();
+      long timestamp = tagHeaderBuffer.readUnsignedInt24();
+      timestamp = (tagHeaderBuffer.readUnsignedByte() << 24) | timestamp;
+      int streamId = tagHeaderBuffer.readUnsignedInt24();
+
+      currentTagHeader.type = type;
+      currentTagHeader.dataSize = dataSize;
+      currentTagHeader.timestamp = timestamp * 1000;
+      currentTagHeader.streamId = streamId;
+
+      // Sanity checks.
+      Assertions.checkState(type == TAG_TYPE_AUDIO || type == TAG_TYPE_VIDEO
+          || type == TAG_TYPE_SCRIPT_DATA);
+      Assertions.checkState(dataSize <= Integer.MAX_VALUE);
+      // Reuse tagData buffer to avoid lot of memory allocation (performance penalty).
+      if (tagData == null || dataSize > tagData.capacity()) {
+        tagData = new ParsableByteArray((int) dataSize);
+      } else {
+        tagData.setPosition(0);
+      }
+      tagData.setLimit(dataSize);
+      parserState = STATE_READING_SAMPLE;
+
+    } catch (EOFException eof) {
+      return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * Reads payload of an FLV tag from the provided {@link ExtractorInput}.
+   * @param input The {@link ExtractorInput} from which to read.
+   * @param seekPosition
+   * @return One of {@link Extractor#RESULT_CONTINUE} and {@link Extractor#RESULT_END_OF_INPUT}.
+   * @throws IOException If an error occurred reading from the source.
+   * @throws InterruptedException If the thread was interrupted.
+   * @throws TagPayloadReader.UnsupportedTrack If payload of the tag is using a codec non
+   * supported codec.
+   */
+  private int readSample(ExtractorInput input, PositionHolder seekPosition) throws IOException,
+      InterruptedException, AudioTagPayloadReader.UnsupportedTrack {
+    if (tagData != null) {
+      if (!input.readFully(tagData.data, 0, currentTagHeader.dataSize, true)) {
+        return RESULT_END_OF_INPUT;
+      }
+      tagData.setPosition(0);
+    } else {
+      input.skipFully(currentTagHeader.dataSize);
+      return RESULT_CONTINUE;
+    }
+
+    // Pass payload to the right payload reader.
+    if (currentTagHeader.type == TAG_TYPE_AUDIO && audioReader != null) {
+      audioReader.consume(tagData, currentTagHeader.timestamp);
+    } else if (currentTagHeader.type == TAG_TYPE_VIDEO && videoReader != null) {
+      videoReader.consume(tagData, currentTagHeader.timestamp);
+    } else if (currentTagHeader.type == TAG_TYPE_SCRIPT_DATA && metadataReader != null) {
+      metadataReader.consume(tagData, currentTagHeader.timestamp);
+      if (metadataReader.durationUs != C.UNKNOWN_TIME_US) {
+        if (audioReader != null) {
+          audioReader.durationUs = metadataReader.durationUs;
```

Comment:
- That's a difficult decision. The problem with waiting for the duration to arrive is there is no any guarantee that duration field is going to arrive at any time. Although that metadata field is typically added at the very beginning of the file, before any other FLV tag (and in fact some CDN's require onMetadata with duration data to be at the very beginning of each FLV file), that's not a must. The only sure way of calculating duration would be reading the last 4 bytes of the FLV file (length of the last FLV tag) and then reading that tag and its PTS value. However, this would require doing an HTTP range request before playback starts that I think modify a bit how Exoplayer HTTP request workflow works. 

For your info, what Flash players are doing is just assuming duration is 0 when metadata duration field is not there (Flash player can't do http byte range requests).

- Ok, makes sense.